network/src/metrics.rs

```diff
@@ -1,155 +1,150 @@
 // Copyright (c) 2022, Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
-use prometheus::{default_registry, register_int_gauge_vec_with_registry, IntGaugeVec, Registry};
+use anemo_tower::callback::MakeCallbackHandler;
+use anemo_tower::callback::ResponseHandler;
+use prometheus::{
+    register_histogram_vec_with_registry, register_int_counter_vec_with_registry,
+    register_int_gauge_vec_with_registry, IntGaugeVec, Registry,
+};
+use prometheus::{HistogramVec, IntCounterVec};
 use std::sync::Arc;
-
-pub trait NetworkMetrics {
-    fn network_available_tasks(&self) -> &IntGaugeVec;
+use std::time::Instant;
+
+#[derive(Clone)]
+pub struct NetworkMetrics1 {
+    /// Counter of requests, route is a label (ie separate timeseries per route)
+    requests: IntCounterVec,
+    /// Request latency, route is a label
+    request_latency: HistogramVec,
+    concurrent_requests: IntGaugeVec,
+    /// Failed requests by route
+    errors: IntCounterVec,
 }
 
-#[derive(Clone, Debug)]
-pub struct PrimaryNetworkMetrics {
-    /// The number of executor available tasks
-    pub network_available_tasks: IntGaugeVec,
-}
+const LATENCY_SEC_BUCKETS: &[f64] = &[
+    0.001, 0.005, 0.01, 0.05, 0.1, 0.25, 0.5, 1., 2.5, 5., 10., 20., 30., 60., 90.,
+];
+
+impl NetworkMetrics1 {
+    pub fn new(node: &'static str, registry: &Registry) -> Self {
+        let requests = register_int_counter_vec_with_registry!(
+            format!("{node}_requests"),
+            "The number of anemo requests made on the network",
+            &["direction", "route"],
```

Comment:
Instead of making direction a label for these metrics, I think I'd sooner have entirely separate metrics for inbound vs outbound? I can't imagine a case where you would want to aggregate these metrics regardless of direction, and in fact it might not even be comparable

For example, with request latency: outbound latency includes the network latency, while inbound latency doesn't. So these are almost like two different units, and it would not only be not-useful but actually incorrect to apply aggregation functions to the union of those

Reply:
I was trying to avoid needing to define two completely different sets of metrics for inbound/outbound for code duplications sake since they'd be easy enough to separate out on the dashboards after the fact